Lib/textwrap.py

```diff
@@ -458,7 +459,14 @@ def dedent(text):
                    "line = %r, margin = %r" % (line, margin)
 
     if margin:
-        text = re.sub(r'(?m)^' + margin, '', text)
+        margin_length = len(margin)
+    else:
+        margin_length = 0
+    text = ''.join([
+        line[margin_length:] if line.lstrip()
+        else line[len(line.splitlines()[0]):]
+        for line in text.splitlines(True)
```

Comment:
Maybe `text.splitlines` could be called once instead of twice ? At the moment, the `keepends` parameter is once False, once True but I guess we could call it with True both times and it wouldn't hurt the `margin` computation.

Also, maybe the `line[foobar:]` could be factorised out as in:

```
text = ''.join([
    line[margin_length if line.lstrip() else len(line.splitlines()[0]):]
    for line in text.splitlines(True)
])
```